Allow localProgress to persist after user Signs up or Logs out

shaavan opened the pull request:
Resolves #416

ecurrencyhodler commented:
Let's get 1 code review. Then I'll QA after we can build this with vercel.

afterburn commented on components/Modals/SignUp.tsx:
```diff
@@ -45,6 +47,7 @@ export default function SignUpModal({ open, onClose }) {
       if (privateKey) {
         await register(privateKey, `/assets/avatars/${avatar}.png`)
         await login(privateKey)
+        saveProgress(progress!)
```
Also I think a comment would help here, I believe this is taking the progress that was stored in LocalStorage and sending that to the backend for storing it in the database right? That is not immediately clear from the code, so in cases like that I think it makes sense to add a comment,

benalleng replied:
where progress could possibly be undefined I think we should just default to setting it to 'CH1INT1'

shaavan replied:
Addressed in b61b1001da3a6ae75a664201d422e40e1384bb32

Thanks!

afterburn reviewed (changes requested):
Small change requested for the non-null assertion.

benalleng commented on app/[lang]/chapters/[slug]/[lesson]/page.tsx:
```diff
@@ -108,18 +108,6 @@ export default function Page({ params }) {
     )
   }
 
-  // If account and progress data have been loaded, but there is no progress obj
-  if (!account || !progress) {
-    return (
-      <>
-        <Head />
-        <div className="flex h-full w-full grow flex-col items-center justify-center">
-          <SignIn lang={params.lang} />
-        </div>
-      </>
-    )
-  }
-
```
Looking back at this I now see what Jonas was saying, didn't we already remove this code? how did it end up back in the codebase from when it was removed in #399

ecurrencyhodler replied:
This was probably a result of us having 6+ PR's pending for a long time all touching various parts of the codebase. @shaavan can you make sure this PR removes this section like you did in #399 again?

shaavan replied:
Addressed in b61b1001da3a6ae75a664201d422e40e1384bb32

Thanks!

afterburn commented on providers/ProgressProvider.tsx:
```diff
@@ -7,7 +7,7 @@ import { getProgressLocal, setProgressLocal } from 'api/local'
 
 export const ProgressContext = createContext<ProgressContextType>({
-  progress: undefined,
+  progress: '',
```
Let's initialize this to `CH1INT1` like @benalleng suggested

afterburn commented on providers/ProgressProvider.tsx:
```diff
@@ -86,7 +86,7 @@ export default function ProgressProvider({
     <ProgressContext.Provider
       value={{
-        progress: accountProgress,
+        progress: accountProgress ?? 'CH1INT1',
```
By initializing the progress variable to `CH1INT1` on line 10, you can probably just do `progress: accountProgress` here.

shaavan replied:
I tried this line of code `progress: accountProgress`, but it causes the following error.

```
Type 'string | undefined' is not assignable to type 'string'.
Type 'undefined' is not assignable to type 'string'.
```

So I can either: 
1. change the `accountProgress` type to string or,
2. Use this (`progress: accountProgress ?? 'CH1INT1'`) maneuver

From these two, I chose the second option because it kept the code clean while being least intrusive.

benalleng replied:
thats so odd, doing some of my own testing, even when setting the useState of account progress to be only a string and defaulted to 'CH1INT1' i still recieve an error on line 46 of `initLocal` with a type error stating that progress cannot be of type `string | undefined` even though it seems to be that progress has type string set everywhere

shaavan replied:
@benalleng 

That's because the progress in the `initLocal` function is equal to the return value of `getProgressLocal` whose return value can be `<string | undefined>`. That's why the error is showing.

![image](https://user-images.githubusercontent.com/85434418/236485289-0be77889-d3c9-4945-a119-dff8e5ae5340.png)

It can be taken as one more reason to not change `accountProgress` type, because of the number of changes it requires to implement this PR's change.

benalleng replied:
I guess my question is why should get ProgressLocal() allow for an undefined type? shouldn't it also default to 'CH1INT1'? 
I guess my thinking is that any instance of progress possibly being undefined would be better handled if it defaulted back to 'CH1INT1' as that is equivalent to base/no progress.

shaavan replied:
This seems to be a better approach to me at first glance.

But I noticed that "undefined" is used in the `getProgressLocal` function when setting res fails for some reason.

```js
export default async function getProgressLocal() {
  try {
    const res = localStorage.getItem('SavingSatoshiProgress') ?? 'CH1INT1'

    return res
  } catch (errors) {
    console.error(errors)
    return undefined
  }
}
```

So what should be the approach you would suggest:

1. Should we replace the undefined with "CH1INT1" even when `getProgressLocal` fails for unspecified reasons?

afterburn reviewed (changes requested):
Thanks @shaavan, minor change requested

benalleng commented on app/[lang]/chapters/[slug]/[lesson]/page.tsx:
```diff
@@ -108,23 +108,11 @@ export default function Page({ params }) {
-  const lastUnlockedLessonPath = getLastUnlockedLessonPath(progress!)
+  const lastUnlockedLessonPath = getLastUnlockedLessonPath(progress)
   const currentLessonPath = `/${pathData.pageId}/${pathData.chapterId}/${pathData.lessonId}`
   const isRestrictedFromLesson = isLessonCompleted(
```
I believe this should be proceeded by a not operand as we are checking that the user has Not completed the specified lesson to determine if they should be redirected.

In testing i am able to type in lesson paths I expect to be locked without being redirected
``` suggestion
  const isRestrictedFromLesson = !isLessonCompleted(
```

shaavan replied:
Here comes @benalleng with his eagle eyes!! 🦅
Great catch! ✨

Since this change fixes a different bug significantly different from the goal of this PR, and since this change is independent of changes done in this PR, it's best to address this change separately to maintain the atomic nature of PR.

Let me create a separate PR for this change!

shaavan replied:
Okay, so this change does depend on changes in this PR (removing the mandatory signup screen)
Let me include this change with this PR.

Thanks again for catching this! 🙌

shaavan replied:
Addressed in 98ad49d932fc391a37a256fa92acfa24b3856fdf

benalleng reviewed (changes requested):
Think a small bug slipped through

shaavan commented:
Updated from b61b1001da3a6ae75a664201d422e40e1384bb32 to 98ad49d932fc391a37a256fa92acfa24b3856fdf ([pr421.02](https://github.com/shaavan/saving-satoshi/commits/pr421.02) -> [pr421.03](https://github.com/shaavan/saving-satoshi/commits/pr421.03), [diff](https://github.com/shaavan/saving-satoshi/compare/pr421.02..pr421.03))
Addressed @benalleng comment.

Changes:

1. Fix the bug related to preventing access to locked challenges.

benalleng reviewed (approved):
tACK for the locked challenges.
local storage progress works as expected.

Testing pathway:
1. Went to "CH1GEN2" via localStorage then logged in, `accountProgress` and `localStorage` was now saved as "CH1GEN2" 
2. proceeded one more lesson to "CH1GEN3" then logged out, local storage remebered "CH1GEN2" and reverted me back one lesson
3. upon logging in with that same private key I was able to return to "CH1GEN3" without needing to use the next button or re-completing the challenge

I am approving as I will defer to @afterburn for the progress and accoutProgress types so that I am not a blocker

shaavan commented:
> tACK for the locked challenges.
local storage progress works as expected.

Yay! 🚀

Let's wait for @afterburn's ACK before merging!

afterburn commented:
Modified the code a bit so that we can initialize the progress provider to `CH1INT1` because this value should never be an empty string, `null` or `undefined`, if no progress can be retrieved for an account we should initialize it to `CH1INT1`. That way, every code that uses `progress` can safely assume that it'll always be a string to run comparisons against.

ecurrencyhodler commented:
tACK for me. Will wait for convo to resolve them merge.
